plotRastersAllUnits.m

```diff
@@ -0,0 +1,121 @@
+function plotRastersAllUnits(nwb_file, trial)
+
+%trial timing details; 
+trial_start_time = nwb_file.intervals_trials.start_time.data(trial);
+% stim
+stim_time = nwb_file.intervals_trials.vectordata.get('visual_stimulus_time').data(trial) - trial_start_time;
+% move
+response_time = nwb_file.intervals_trials.vectordata.get('response_time').data(trial) - trial_start_time;
+% tone
+tone_time = nwb_file.intervals_trials.vectordata.get('go_cue').data(trial) - trial_start_time;
+% reward
+feedback_time = nwb_file.intervals_trials.vectordata.get('feedback_time').data(trial) - trial_start_time;
+
+% time details for raster plots
+before_time = 0;
+after_time = 1 +feedback_time;
+
+%%% Electrode details
+probe_list = keys(nwb_file.general_extracellular_ephys.map);
+nprobes = length(probe_list);
+% electrode location
+electrode_area = nwb_file.general_extracellular_ephys_electrodes.vectordata.get('location').data(:);
+
+%%% All unit/cluster details
+% get cluster annotations
+annot = nwb_file.units.vectordata.get('phy_annotations').data(:);
+% exclude MUA and noise (as in Steinmetz et al 2019 paper)
+valid_units = find(annot>=2);
+num_units = length(valid_units);
+% recording probe
+probe_all = arrayfun( ...
+    @(x) probePathToIdx(x.path, probe_list), ...
+    nwb_file.units.electrode_group.data ...
+);
+% cluster channel
+channel_all = nwb_file.units.vectordata.get('peak_channel').data(:);
+% cluster depth (0 = deepest, 3820 = shallowest; scale um)
+depth_all = nwb_file.units.vectordata.get('cluster_depths').data(:);
+
+% get details for valid units
+depth_unit = depth_all(valid_units);
+probe_unit = probe_all(valid_units);
+area_unit = electrode_area(channel_all(valid_units));
+
+% get raster data
+raster_data = {};
+for n = 1:num_units
+    % get spike times
+    spike_times = util.read_indexed_column(nwb_file.units.spike_times_index, ...
+                                           nwb_file.units.spike_times, ...
+                                           valid_units(n));
+    raster_data{end+1} = spike_times(...
+                       spike_times >= trial_start_time + before_time & ...
+                       spike_times <= trial_start_time + after_time) - ...
+                       trial_start_time;
+end
```

Comment:
currently takes ~25 secs to output the figure with this function. 20 secs are spent in this for loop. I'm gonna try to modify things a bit to get better performance so that live script can have less lag between selection of trial and producing the figure.

Reply:
modifying this bit to minimize I/O greatly improved performance.